include/aws/common/task_scheduler.h

```diff
@@ -0,0 +1,101 @@
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/common.h>
+#include <aws/common/priority_queue.h>
+#include <stdint.h>
+
+typedef void *aws_task_arg;
+
+/**
+ * Pointer to the scheduled function
+ */
+typedef void(*aws_task_fn)(aws_task_arg arg);
+
+/**
+ * Monotonic clock function
+ */
+typedef int(*aws_task_scheduler_clock)(uint64_t *);
+
+struct aws_task {
+    aws_task_fn fn;
+    aws_task_arg arg;
+};
+
+struct aws_task_scheduler {
+    struct aws_allocator *alloc;
+    struct aws_priority_queue queue;
+    uint64_t owning_thread;
+    aws_task_scheduler_clock clock;
+};
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+    /**
+     * Initializes a task scheduler instance with a clock function and config. A priority queue will be allocated 
+     * in dynamic mode with initial size of 7
+     * The recommended clock functions are in <aws/common/clock.h>
+     *
+     * Note: This needs to be called from the thread that will own the scheduler. Cross-thread access is not allowed 
+     * and will always result in an error code being returned.
+     */
+    AWS_COMMON_API int aws_task_scheduler_init_default(struct aws_task_scheduler *scheduler, 
+            aws_task_scheduler_clock clock, struct aws_allocator *alloc);
+
+    /**
+     * Initializes a task scheduler instance with a priority queue, clock function and config.
+     * The recommended clock functions are in <aws/common/clock.h>
+     *
+     * Note: This needs to be called from the thread that will own the scheduler. Cross-thread access is not allowed 
+     * and will always result in an error code being returned.
+     */
+    AWS_COMMON_API void aws_task_scheduler_init_with_queue(struct aws_task_scheduler *scheduler, 
+            struct aws_priority_queue *queue, aws_task_scheduler_clock clock);
+
+    /**
+     * Cleans up any memory allocated, and prepares the instance for reuse or deletion.
+     */
+    AWS_COMMON_API void aws_task_scheduler_clean_up(struct aws_task_scheduler *scheduler);
+
+    /**
+     * Gets the next task that is ready for execution. Tasks in the future will not be returned, 
+     * even if they are the highest priority. next_run_time, if it is not null, and there are any tasks scheduled
+     * will be set to the timestamp for the highest priority task. This is a useful hint for setting timeouts on
+     * event loops, or thread sleeps. If no tasks are scheduled, this value will be set to 0. task is copied
+     */
+    AWS_COMMON_API int aws_task_scheduler_next_task(struct aws_task_scheduler *scheduler, struct aws_task *task, 
+            uint64_t *next_run_time);
+
+    /**
+     * Schedules a task to run at whatever the current return value of the clock fn is. task is copied
+     */
+    AWS_COMMON_API int aws_task_scheduler_schedule_now(struct aws_task_scheduler *queue, struct aws_task *task);
+
+    /**
+     * Schedules a task to run at time_to_run relative to the clock fn. task is copied
```

Comment:
Don't talk about the clock fn in every function. Just mention in the initialize calls that it is used to query for the current time, and explain elsewhere that e.g. `task is queued to execute time_to_run time units after the current time`